File: packages/crewai-agent/src/on_brand_crew/tools/brand_checker.py
"""Brand checking tools for CrewAI agents."""

import hashlib
import json
from datetime import datetime
from pathlib import Path
from typing import Any, Optional, Type

from crewai.tools import BaseTool
from pydantic import BaseModel, Field

from on_brand_crew.models import (
    BrandProfile,
    BrandCheckRequest,
    BrandCheckResponse,
    BrandAlignmentStatus,
    BrandExplanation,
    ExplanationSeverity,
    ExplanationAspect,
)


def compute_content_hash(content: str) -> str:
    """Compute SHA-256 hash of content."""
    return hashlib.sha256(content.encode()).hexdigest()


def normalize_text(text: str) -> str:
    """Normalize text for comparison."""
    import re
    return re.sub(r'\s+', ' ', text.lower().strip())


def contains_phrase(content: str, phrase: str) -> bool:
    """Check if content contains a phrase (case-insensitive)."""
    return normalize_text(phrase) in normalize_text(content)


def tokenize(text: str) -> set[str]:
    """Tokenize text into words."""
    import re
    normalized = normalize_text(text)
    words = re.split(r'[\s.,!?;:\'"()\[\]{}]+', normalized)
    return {w for w in words if w}


def calculate_word_overlap(text1: str, text2: str) -> float:
    """Calculate Jaccard similarity between two texts."""
    words1 = tokenize(text1)
    words2 = tokenize(text2)
    if not words1 or not words2:
        return 0.0
    intersection = len(words1 & words2)
    union = len(words1 | words2)
    return intersection / union if union > 0 else 0.0


class LoadBrandProfileInput(BaseModel):
    """Input for loading a brand profile."""
    profile_path: str = Field(
        description="Path to the brand profile JSON file"
    )


class LoadBrandProfileTool(BaseTool):
    """Tool to load a brand profile from a JSON file."""

    name: str = "load_brand_profile"
    description: str = (
        "Load a brand profile from a JSON file. "
        "Returns the parsed brand profile with values, voice descriptors, "
        "tone boundaries, never rules, and examples."
    )
    args_schema: Type[BaseModel] = LoadBrandProfileInput

    def _run(self, profile_path: str) -> dict[str, Any]:
        """Load and validate a brand profile."""
        path = Path(profile_path)
        if not path.exists():
            return {"error": f"Profile not found: {profile_path}"}

        try:
            with open(path) as f:
                data = json.load(f)
            profile = BrandProfile.model_validate(data)
            return profile.model_dump(by_alias=True)
        except json.JSONDecodeError as e:
            return {"error": f"Invalid JSON: {e}"}
        except Exception as e:
            return {"error": f"Failed to load profile: {e}"}


class CheckNeverRulesInput(BaseModel):
    """Input for checking never rules."""
    content: str = Field(description="The content to check")
    never_rules: list[str] = Field(description="List of never rules to check against")


class CheckNeverRulesTool(BaseTool):
    """Tool to check content against 'never do/say' rules."""

    name: str = "check_never_rules"
    description: str = (
        "Check if content violates any 'never do/say' rules. "
        "Returns a list of violated rules and whether the content passes."
    )
    args_schema: Type[BaseModel] = CheckNeverRulesInput

    def _run(self, content: str, never_rules: list[str]) -> dict[str, Any]:
        """Check content against never rules."""
        violated = []
        passed = []

        for rule in never_rules:
            if contains_phrase(content, rule):
                violated.append(rule)
            else:
                passed.append(rule)

        return {
            "violated": sorted(violated),
            "passed": sorted(passed),
            "has_violations": len(violated) > 0,
            "violation_count": len(violated),
        }


class CheckToneInput(BaseModel):
    """Input for checking tone boundaries."""
    content: str = Field(description="The content to check")
    acceptable: list[str] = Field(description="List of acceptable tone characteristics")
    unacceptable: list[str] = Field(description="List of unacceptable tone characteristics")


class CheckToneTool(BaseTool):
    """Tool to check content against tone boundaries."""

    name: str = "check_tone"
    description: str = (
        "Check if content matches acceptable tone and avoids unacceptable tone. "
        "Returns found acceptable and unacceptable tone characteristics."
    )
    args_schema: Type[BaseModel] = CheckToneInput

    def _run(
        self,
        content: str,
        acceptable: list[str],
        unacceptable: list[str]
    ) -> dict[str, Any]:
        """Check content against tone boundaries."""
        acceptable_found = []
        unacceptable_found = []

        for tone in acceptable:
            if contains_phrase(content, tone):
                acceptable_found.append(tone)

        for tone in unacceptable:
            if contains_phrase(content, tone):
                unacceptable_found.append(tone)

        return {
            "acceptable_found": sorted(acceptable_found),
            "unacceptable_found": sorted(unacceptable_found),
            "has_violations": len(unacceptable_found) > 0,
            "tone_score": len(acceptable_found) / max(len(acceptable), 1),
        }


class CheckValueAlignmentInput(BaseModel):
    """Input for checking value alignment."""
    content: str = Field(description="The content to check")
    values: list[str] = Field(description="List of brand values")
    voice_descriptors: list[str] = Field(description="List of voice descriptors")


class CheckValueAlignmentTool(BaseTool):
    """Tool to check content alignment with brand values and voice."""

    name: str = "check_value_alignment"
    description: str = (
        "Check how well content aligns with brand values and voice descriptors. "
        "Returns alignment scores and matched/missing elements."
    )
    args_schema: Type[BaseModel] = CheckValueAlignmentInput

    def _run(
        self,
        content: str,
        values: list[str],
        voice_descriptors: list[str]
    ) -> dict[str, Any]:
        """Check content alignment with values and voice."""
        # Check values
        values_aligned = []
        values_missing = []
        for value in values:
            if contains_phrase(content, value):
                values_aligned.append(value)
            else:
                values_missing.append(value)

        # Check voice
        voice_aligned = []
        voice_missing = []
        for descriptor in voice_descriptors:
            if contains_phrase(content, descriptor):
                voice_aligned.append(descriptor)
            else:
                voice_missing.append(descriptor)

        value_score = len(values_aligned) / max(len(values), 1)
        voice_score = len(voice_aligned) / max(len(voice_descriptors), 1)

        return {
            "values_aligned": sorted(values_aligned),
            "values_missing": sorted(values_missing),
            "voice_aligned": sorted(voice_aligned),
            "voice_missing": sorted(voice_missing),
            "value_score": value_score,
            "voice_score": voice_score,
            "combined_score": (value_score + voice_score) / 2,
        }


class BrandCheckerInput(BaseModel):
    """Input for the complete brand check."""
    content: str = Field(description="The content to check for brand consistency")
    profile_path: str = Field(
        description="Path to the brand profile JSON file",
        default="./brand-profile.json"
    )
    content_type: Optional[str] = Field(
        None,
        description="Type of content (ad-copy, social-post, etc.)"
    )


class BrandCheckerTool(BaseTool):
    """Complete brand consistency checking tool."""

    name: str = "check_brand_consistency"
    description: str = (
        "Perform a complete brand consistency check on content. "
        "Evaluates content against the brand profile including values, voice, "
        "tone, never rules, and examples. Returns status (on-brand, borderline, "
        "off-brand), explanations, and confidence score."
    )
    args_schema: Type[BaseModel] = BrandCheckerInput

    def _run(
        self,
        content: str,
        profile_path: str = "./brand-profile.json",
        content_type: Optional[str] = None
    ) -> dict[str, Any]:
        """Perform complete brand consistency check."""
        # Load profile
        path = Path(profile_path)
        if not path.exists():
            return {"error": f"Profile not found: {profile_path}"}

        try:
            with open(path) as f:
                data = json.load(f)
            profile = BrandProfile.model_validate(data)
        except Exception as e:
            return {"error": f"Failed to load profile: {e}"}

        explanations: list[dict[str, Any]] = []
        status = BrandAlignmentStatus.ON_BRAND
        confidence = 85

        # 1. Check never rules (critical)
        never_violations = []
        for rule in profile.never_rules:
            if contains_phrase(content, rule):
                never_violations.append(rule)

        if never_violations:
            status = BrandAlignmentStatus.OFF_BRAND
            confidence = 95
            explanations.append({
                "text": f'Contains prohibited content: "{never_violations[0]}"',
                "aspect": "never-rule",
                "severity": "critical",
            })

        # 2. Check tone boundaries
        unacceptable_found = []
        acceptable_found = []
        for tone in profile.tone_unacceptable:
            if contains_phrase(content, tone):
                unacceptable_found.append(tone)
        for tone in profile.tone_acceptable:
            if contains_phrase(content, tone):
                acceptable_found.append(tone)

        if unacceptable_found:
            status = BrandAlignmentStatus.OFF_BRAND
            confidence = max(confidence, 90)
            if len(explanations) < 3:
                explanations.append({
                    "text": f'Uses unacceptable tone: "{unacceptable_found[0]}"',
                    "aspect": "tone",
                    "severity": "critical",
                })

        # 3. Check example similarity
        good_similarity = 0.0
        bad_similarity = 0.0
        for example in profile.examples:
            sim = calculate_word_overlap(content, example.content)
            if example.type == "good":
                good_similarity = max(good_similarity, sim)
            else:
                bad_similarity = max(bad_similarity, sim)

        if bad_similarity > 0.3:
            if status != BrandAlignmentStatus.OFF_BRAND:
                status = (
                    BrandAlignmentStatus.OFF_BRAND
                    if bad_similarity > 0.5
                    else BrandAlignmentStatus.BORDERLINE
                )
            confidence = max(confidence, 80)
            if len(explanations) < 3:
                explanations.append({
                    "text": "Content resembles known off-brand examples",
                    "aspect": "example-match",
                    "severity": "critical" if bad_similarity > 0.5 else "warning",
                })

        # 4. Check value and voice alignment
        values_aligned = [v for v in profile.values if contains_phrase(content, v)]
        voice_aligned = [v for v in profile.voice_descriptors if contains_phrase(content, v)]

        value_score = len(values_aligned) / max(len(profile.values), 1)
        voice_score = len(voice_aligned) / max(len(profile.voice_descriptors), 1)
        combined_score = (value_score + voice_score) / 2

        if status == BrandAlignmentStatus.ON_BRAND:
            if combined_score < 0.3:
                status = BrandAlignmentStatus.BORDERLINE
                confidence = 70
            elif combined_score < 0.5:
                status = BrandAlignmentStatus.BORDERLINE
                confidence = 75
            else:
                confidence = 80 + int(combined_score * 15)

        # Add positive/constructive explanations
        if not explanations:
            if good_similarity > 0.3:
                explanations.append({
                    "text": "Content aligns well with established brand examples",
                    "aspect": "example-match",
                    "severity": "info",
                })
            elif value_score > 0.5:
                aligned_str = ", ".join(values_aligned[:2])
                explanations.append({
                    "text": f"Content reflects brand values: {aligned_str}",
                    "aspect": "value",
                    "severity": "info",
                })
            else:
                explanations.append({
                    "text": "Content is acceptable but could better reflect brand values",
                    "aspect": "value",
                    "severity": "info",
                })

        # Add voice explanation if space
        if len(explanations) < 3 and voice_score < 0.5 and status != BrandAlignmentStatus.OFF_BRAND:
            voice_missing = [v for v in profile.voice_descriptors if not contains_phrase(content, v)]
            if voice_missing:
                missing_str = ", ".join(voice_missing[:2])
                explanations.append({
                    "text": f"Voice could better emphasize: {missing_str}",
                    "aspect": "voice",
                    "severity": "info",
                })

        # Add acceptable tone if space
        if len(explanations) < 3 and acceptable_found:
            found_str = ", ".join(acceptable_found[:2])
            explanations.append({
                "text": f"Good use of brand tone: {found_str}",
                "aspect": "tone",
                "severity": "info",
            })

        # Ensure at least one explanation
        if not explanations:
            explanations.append({
                "text": (
                    "Content aligns with brand guidelines"
                    if status == BrandAlignmentStatus.ON_BRAND
                    else "Content requires review for brand alignment"
                ),
                "severity": "info",
            })

        # Sort by severity
        severity_order = {"critical": 0, "warning": 1, "info": 2}
        explanations.sort(key=lambda x: severity_order.get(x.get("severity", "info"), 2))
        explanations = explanations[:3]

        # Build status display
        status_display = {
            BrandAlignmentStatus.ON_BRAND: "On Brand ✅",
            BrandAlignmentStatus.BORDERLINE: "Borderline ⚠️",
            BrandAlignmentStatus.OFF_BRAND: "Off Brand ❌",
        }[status]

        return {
            "status": status.value,
            "statusDisplay": status_display,
            "explanations": explanations,
            "confidence": confidence,
            "profileVersion": profile.version,
            "checkedAt": datetime.utcnow().isoformat() + "Z",
            "contentHash": compute_content_hash(content),
            "details": {
                "neverRuleViolations": sorted(never_violations),
                "unacceptableToneFound": sorted(unacceptable_found),
                "valueAlignmentScore": value_score,
                "voiceAlignmentScore": voice_score,
                "exampleSimilarity": {
                    "good": good_similarity,
                    "bad": bad_similarity,
                },
            },
        }

File: packages/core/src/brand-checker.ts
import { createHash } from 'crypto';
import {
  BrandProfile,
  BrandCheckRequest,
  BrandCheckResponse,
  BrandAlignmentStatus,
  BrandExplanation,
  BrandCheckAuditEntry,
} from '@agent-resolver/schema';

/**
 * Compute deterministic hash for content
 */
export function computeContentHash(content: string): string {
  return createHash('sha256').update(content).digest('hex');
}

/**
 * Generate a unique audit ID
 */
export function generateAuditId(contentHash: string, timestamp: string): string {
  return createHash('sha256')
    .update(`${contentHash}|${timestamp}`)
    .digest('hex')
    .substring(0, 16);
}

/**
 * Normalize text for comparison (lowercase, trim, collapse whitespace)
 */
function normalizeText(text: string): string {
  return text.toLowerCase().trim().replace(/\s+/g, ' ');
}

/**
 * Tokenize text into words
 */
function tokenize(text: string): string[] {
  return normalizeText(text)
    .split(/[\s.,!?;:'"()\[\]{}]+/)
    .filter((word) => word.length > 0);
}

/**
 * Check if content contains a phrase (case-insensitive)
 */
function containsPhrase(content: string, phrase: string): boolean {
  const normalizedContent = normalizeText(content);
  const normalizedPhrase = normalizeText(phrase);
  return normalizedContent.includes(normalizedPhrase);
}

/**
 * Calculate word overlap score between two texts
 */
function calculateWordOverlap(text1: string, text2: string): number {
  const words1 = new Set(tokenize(text1));
  const words2 = new Set(tokenize(text2));

  if (words1.size === 0 || words2.size === 0) {
    return 0;
  }

  let overlap = 0;
  for (const word of words1) {
    if (words2.has(word)) {
      overlap++;
    }
  }

  // Jaccard similarity
  const union = new Set([...words1, ...words2]).size;
  return overlap / union;
}

/**
 * Check content against "never" rules
 */
function checkNeverRules(
  content: string,
  neverRules: readonly string[]
): { violated: string[]; passed: string[] } {
  const violated: string[] = [];
  const passed: string[] = [];

  for (const rule of neverRules) {
    if (containsPhrase(content, rule)) {
      violated.push(rule);
    } else {
      passed.push(rule);
    }
  }

  // Sort for determinism
  violated.sort();
  passed.sort();

  return { violated, passed };
}

/**
 * Check content against voice descriptors
 */
function checkVoiceAlignment(
  content: string,
  voiceDescriptors: readonly string[]
): { aligned: string[]; missing: string[] } {
  const aligned: string[] = [];
  const missing: string[] = [];

  for (const descriptor of voiceDescriptors) {
    // Simple keyword check - in production this would use embeddings
    if (containsPhrase(content, descriptor)) {
      aligned.push(descriptor);
    } else {
      missing.push(descriptor);
    }
  }

  // Sort for determinism
  aligned.sort();
  missing.sort();

  return { aligned, missing };
}

/**
 * Check content against tone boundaries
 */
function checkToneBoundaries(
  content: string,
  acceptable: readonly string[],
  unacceptable: readonly string[]
): { acceptableFound: string[]; unacceptableFound: string[] } {
  const acceptableFound: string[] = [];
  const unacceptableFound: string[] = [];

  for (const tone of acceptable) {
    if (containsPhrase(content, tone)) {
      acceptableFound.push(tone);
    }
  }

  for (const tone of unacceptable) {
    if (containsPhrase(content, tone)) {
      unacceptableFound.push(tone);
    }
  }

  // Sort for determinism
  acceptableFound.sort();
  unacceptableFound.sort();

  return { acceptableFound, unacceptableFound };
}

/**
 * Check content against brand values
 */
function checkValueAlignment(
  content: string,
  values: readonly string[]
): { aligned: string[]; missing: string[] } {
  const aligned: string[] = [];
  const missing: string[] = [];

  for (const value of values) {
    if (containsPhrase(content, value)) {
      aligned.push(value);
    } else {
      missing.push(value);
    }
  }

  // Sort for determinism
  aligned.sort();
  missing.sort();

  return { aligned, missing };
}

/**
 * Calculate similarity with canonical examples
 */
function checkExampleSimilarity(
  content: string,
  examples: BrandProfile['examples']
): { goodSimilarity: number; badSimilarity: number; bestMatch: string | null } {
  const goodExamples = examples.filter((e) => e.type === 'good');
  const badExamples = examples.filter((e) => e.type === 'bad');

  let goodSimilarity = 0;
  let badSimilarity = 0;
  let bestMatch: string | null = null;
  let bestScore = 0;

  for (const example of goodExamples) {
    const score = calculateWordOverlap(content, example.content);
    if (score > goodSimilarity) {
      goodSimilarity = score;
    }
    if (score > bestScore) {
      bestScore = score;
      bestMatch = example.content.substring(0, 50) + (example.content.length > 50 ? '...' : '');
    }
  }

  for (const example of badExamples) {
    const score = calculateWordOverlap(content, example.content);
    if (score > badSimilarity) {
      badSimilarity = score;
    }
  }

  return { goodSimilarity, badSimilarity, bestMatch };
}

/**
 * Brand check result with internal scoring
 */
interface BrandCheckResult {
  status: BrandAlignmentStatus;
  explanations: BrandExplanation[];
  confidence: number;
  details: {
    neverRuleViolations: string[];
    unacceptableToneFound: string[];
    valueAlignmentScore: number;
    voiceAlignmentScore: number;
    exampleSimilarity: { good: number; bad: number };
  };
}

/**
 * Core brand consistency check - deterministic logic
 *
 * This function evaluates content against a brand profile and returns
 * a structured result with status, explanations, and confidence score.
 *
 * The algorithm:
 * 1. Check for "never rule" violations (critical)
 * 2. Check for unacceptable tone usage (critical)
 * 3. Check similarity with bad examples (warning)
 * 4. Check value alignment (scoring)
 * 5. Check voice alignment (scoring)
 * 6. Check similarity with good examples (scoring)
 *
 * Status determination:
 * - OFF-BRAND: Any critical violation (never rules, unacceptable tone, high bad example similarity)
 * - BORDERLINE: Low scores or warnings without critical violations
 * - ON-BRAND: Good alignment without violations
 */
export function checkBrandConsistency(
  profile: BrandProfile,
  request: BrandCheckRequest
): BrandCheckResult {
  const content = request.content;
  const explanations: BrandExplanation[] = [];

  // 1. Check never rules (critical)
  const neverCheck = checkNeverRules(content, profile.neverRules);

  // 2. Check tone boundaries
  const toneCheck = checkToneBoundaries(
    content,
    profile.toneAcceptable,
    profile.toneUnacceptable
  );

  // 3. Check example similarity
  const exampleCheck = checkExampleSimilarity(content, profile.examples);

  // 4. Check value alignment
  const valueCheck = checkValueAlignment(content, profile.values);
  const valueScore = profile.values.length > 0
    ? valueCheck.aligned.length / profile.values.length
    : 1;

  // 5. Check voice alignment
  const voiceCheck = checkVoiceAlignment(content, profile.voiceDescriptors);
  const voiceScore = profile.voiceDescriptors.length > 0
    ? voiceCheck.aligned.length / profile.voiceDescriptors.length
    : 1;

  // Determine status and build explanations
  let status: BrandAlignmentStatus = 'on-brand';
  let confidence = 85; // Base confidence

  // Critical: Never rule violations
  if (neverCheck.violated.length > 0) {
    status = 'off-brand';
    confidence = 95;
    explanations.push({
      text: `Contains prohibited content: "${neverCheck.violated[0]}"`,
      aspect: 'never-rule',
      severity: 'critical',
    });
  }

  // Critical: Unacceptable tone
  if (toneCheck.unacceptableFound.length > 0) {
    status = 'off-brand';
    confidence = Math.max(confidence, 90);
    if (explanations.length < 3) {
      explanations.push({
        text: `Uses unacceptable tone: "${toneCheck.unacceptableFound[0]}"`,
        aspect: 'tone',
        severity: 'critical',
      });
    }
  }

  // Warning: High similarity to bad examples
  if (exampleCheck.badSimilarity > 0.3) {
    if (status !== 'off-brand') {
      status = exampleCheck.badSimilarity > 0.5 ? 'off-brand' : 'borderline';
    }
    confidence = Math.max(confidence, 80);
    if (explanations.length < 3) {
      explanations.push({
        text: `Content resembles known off-brand examples`,
        aspect: 'example-match',
        severity: exampleCheck.badSimilarity > 0.5 ? 'critical' : 'warning',
      });
    }
  }

  // Scoring: Check alignment levels
  if (status === 'on-brand') {
    const combinedScore = (valueScore + voiceScore) / 2;

    if (combinedScore < 0.3) {
      status = 'borderline';
      confidence = 70;
    } else if (combinedScore < 0.5) {
      status = 'borderline';
      confidence = 75;
    } else {
      confidence = 80 + Math.round(combinedScore * 15);
    }
  }

  // Add positive or constructive explanations
  if (explanations.length === 0) {
    if (exampleCheck.goodSimilarity > 0.3) {
      explanations.push({
        text: `Content aligns well with established brand examples`,
        aspect: 'example-match',
        severity: 'info',
      });
    } else if (valueScore > 0.5) {
      explanations.push({
        text: `Content reflects brand values: ${valueCheck.aligned.slice(0, 2).join(', ')}`,
        aspect: 'value',
        severity: 'info',
      });
    } else {
      explanations.push({
        text: `Content is acceptable but could better reflect brand values`,
        aspect: 'value',
        severity: 'info',
      });
    }
  }

  // Add voice alignment explanation if space
  if (explanations.length < 3 && voiceScore < 0.5 && status !== 'off-brand') {
    const missingVoice = voiceCheck.missing.slice(0, 2).join(', ');
    if (missingVoice) {
      explanations.push({
        text: `Voice could better emphasize: ${missingVoice}`,
        aspect: 'voice',
        severity: 'info',
      });
    }
  }

  // Add acceptable tone explanation if space
  if (explanations.length < 3 && toneCheck.acceptableFound.length > 0) {
    explanations.push({
      text: `Good use of brand tone: ${toneCheck.acceptableFound.slice(0, 2).join(', ')}`,
      aspect: 'tone',
      severity: 'info',
    });
  }

  // Ensure we have at least one explanation
  if (explanations.length === 0) {
    explanations.push({
      text: status === 'on-brand'
        ? 'Content aligns with brand guidelines'
        : 'Content requires review for brand alignment',
      severity: 'info',
    });
  }

  // Sort explanations by severity for determinism
  explanations.sort((a, b) => {
    const severityOrder = { critical: 0, warning: 1, info: 2 };
    return severityOrder[a.severity] - severityOrder[b.severity];
  });

  // Limit to 3 explanations
  const finalExplanations = explanations.slice(0, 3);

  return {
    status,
    explanations: finalExplanations,
    confidence,
    details: {
      neverRuleViolations: neverCheck.violated,
      unacceptableToneFound: toneCheck.unacceptableFound,
      valueAlignmentScore: valueScore,
      voiceAlignmentScore: voiceScore,
      exampleSimilarity: {
        good: exampleCheck.goodSimilarity,
        bad: exampleCheck.badSimilarity,
      },
    },
  };
}

/**
 * Format status with emoji for display
 */
export function formatStatusDisplay(status: BrandAlignmentStatus): string {
  switch (status) {
    case 'on-brand':
      return 'On Brand ✅';
    case 'borderline':
      return 'Borderline ⚠️';
    case 'off-brand':
      return 'Off Brand ❌';
  }
}

/**
 * Create a full brand check response
 */
export function createBrandCheckResponse(
  profile: BrandProfile,
  request: BrandCheckRequest,
  timestamp?: string
): BrandCheckResponse {
  const result = checkBrandConsistency(profile, request);
  const checkedAt = timestamp ?? new Date().toISOString();
  const contentHash = computeContentHash(request.content);

  return {
    status: result.status,
    statusDisplay: formatStatusDisplay(result.status),
    explanations: result.explanations,
    confidence: result.confidence,
    profileVersion: profile.version,
    checkedAt,
    contentHash,
  };
}

/**
 * Create an audit log entry from a brand check response
 */
export function createAuditEntry(
  profile: BrandProfile,
  response: BrandCheckResponse
): BrandCheckAuditEntry {
  return {
    id: generateAuditId(response.contentHash, response.checkedAt),
    timestamp: response.checkedAt,
    profileName: profile.name,
    profileVersion: response.profileVersion,
    contentHash: response.contentHash,
    status: response.status,
    confidence: response.confidence,
  };
}

/**
 * Validate a brand profile
 */
export function validateBrandProfile(
  data: unknown
): { success: true; data: BrandProfile } | { success: false; errors: string[] } {
  const result = BrandProfile.safeParse(data);
  if (result.success) {
    return { success: true, data: result.data };
  }
  return {
    success: false,
    errors: result.error.errors.map(
      (e) => `${e.path.join('.')}: ${e.message}`
    ),
  };
}

/**
 * Validate a brand check request
 */
export function validateBrandCheckRequest(
  data: unknown
): { success: true; data: BrandCheckRequest } | { success: false; errors: string[] } {
  const result = BrandCheckRequest.safeParse(data);
  if (result.success) {
    return { success: true, data: result.data };
  }
  return {
    success: false,
    errors: result.error.errors.map(
      (e) => `${e.path.join('.')}: ${e.message}`
    ),
  };
}
